subiquity/common/types.py

```diff
@@ -341,10 +341,14 @@ class StorageResponse:
 
 @attr.s(auto_attribs=True)
 class StorageResponseV2:
-    disks: List[Disk]
-    need_root: bool  # if True, there is not yet a partition mounted at "/"
-    need_boot: bool  # if True, there is not yet a boot partition
-    install_minimum_size: int
+    status: ProbeStatus
+    error_report: Optional[ErrorReportRef] = None
+    disks: List[Disk] = attr.Factory(list)
+    # if need_root == True, there is not yet a partition mounted at "/"
+    need_root: Optional[bool] = None
+    # if need_boot == True, there is not yet a boot partition
+    need_boot: Optional[bool] = None
+    install_minimum_size: Optional[int] = None
```

Comment:
I'd usually tend to wrap the fields that depend on `status` in an object itself marked optional, as in:

```python
@attr.s(auto_attribs=True)
class StorageResponseDoneV2:
    disks: List[Disk]
    need_root: bool  # if True, there is not yet a partition mounted at "/"
    need_boot: bool  # if True, there is not yet a boot partition
    install_minimum_size: int

@attr.s(auto_attribs=True)
class StorageResponseV2:
    status: ProbeStatus
    probe_info: Optional[StorageResponseDoneV2]
    error_report: Optional[ErrorReportRef] = None
```
would it make sense here? I know it breaks backward compatibility for the desktop team but I'm sure it's not a problem since they will add the `wait=True` parameter to their existing invocations anyway.

Reply:
Yes, adding two elements to several return types feels a bit awkward.
I think I want to reverse that, and keep the main data the the caller is interested in at the top level.
Something like:
```
@attr.s(auto_attribs=True)
class ProbeResult:
    status: ProbeStatus
    error_report: Optional[ErrorReportRef] = None

@attr.s(auto_attribs=True)
class StorageResponseV2:
    result: ProbeResult
    disks: Optional[List[Disk]]
    ...
```
Then `_probe_response` can just return the result object, and the caller can handle the rest (and even partially populate the parent object instead of using defaults).
Will add to my future-PR list.